File: python/character_match.py

def is_character_match(string1, string2):

    result = False

    # Convert to lower, split at into list at the space character
    first_string_list = string1.lower().split(" ")
    second_string_list = string2.lower().split(" ")

    # Rejoin the strings without spaces.
    first_string = "".join(first_string_list)
    second_string = "".join(second_string_list)

    # If string lengths are different, they can't be anagrams, return false
    if len(first_string) != len(second_string):
        return result

    string_dict = {}

    for i in range(len(first_string)):

        if first_string[i] in string_dict:
            string_dict[first_string[i]] += 1
        else:
            string_dict[first_string[i]] = 1

    for i in range (len(string_dict)):
        #Use the .get() to handle if character:count pair is not in string_dict.
        if(string_dict.get(second_string[i])):
            string_dict[second_string[i]] -= 1
        else:
            return result
    
    result = True
    return result

File: js/characterMatch.js
exports.isCharacterMatch = function(string1, string2) {
  /**
   * This program takes in two strings and compares them
   * Case doesn't matter, returns true if the first string can be rearranged to be the second, false otherwise.
   * Catch the edge cases:
   * str1.length > || < str2.length
   */
  let result = false
  // Convert input strings to lower case, and discard spaces.
  firstString = string1.toLowerCase()
  .split(" ").join("");
  secondString = string2.toLowerCase()
  .split(" ").join("");

  if(firstString.length !== secondString.length) return result
  let string1Chars = {};
  for (let i = 0; i<firstString.length; i++) {
    string1Chars[firstString[i]] = (string1Chars[firstString[i]] || 0) + 1
  }

  for (i = 0; i < firstString.length; i++) {
    if(string1Chars[secondString[i]]) {
      //Decrement the Object Letter:Count pair by one while iterating through the second string
      string1Chars[secondString[i]]--
    } else return result
  }
  result = true
  return result
};